refactor: separate graph from image

dmcgowan opened the pull request:
Move graph related functions in image to graph package.
Consolidating graph functionality is the first step in refactoring graph into an image store model.
Subsequent refactors will involve breaking up graph into multiple types with a strongly defined interface.

stevvooe commented on graph/graph.go:
```diff
@@ -428,10 +427,125 @@ func (graph *Graph) Heads() (map[string]*image.Image, error) {
+}
+
+// SetCheckSum sets the checksum for the image layer to the provided value.
+func (graph *Graph) SetCheckSum(id, checksum string) error {
```
Do we want to make this typesafe with digest.Digest?

dmcgowan replied:
Hadn't considered it, but I will explore

dmcgowan replied:
GetCheckSum should also be made type safe as well. If it is expected GetCheckSum will return a valid Digest, then it will also need to return an error when there is no CheckSum.

stevvooe commented on graph/graph.go:
```diff
@@ -428,10 +427,125 @@ func (graph *Graph) Heads() (map[string]*image.Image, error) {
+}
+
+// RawJson returns the JSON representation for an image as a byte array.
+func (graph *Graph) RawJson(img *image.Image) ([]byte, error) {
```
`RawJSON`.

stevvooe replied:
Oh, actually, does the serialization of an image belong on the `Graph` object? For example, is it related to how the graph stores the image? It may be but perhaps we could make this relationship more clear.

dmcgowan replied:
This function is strange since it only uses the image ID and pulls the contents directly from disk. The best move is probably to just take in the `id` instead of the entire `*image.Image`

stevvooe replied:
That might be better.

stevvooe commented on graph/graph_test.go:
```diff
@@ -301,6 +301,6 @@ func tempGraph(t *testing.T) (*Graph, graphdriver.Driver) {
 }
 
 func nukeGraph(graph *Graph) {
```
`func (g *Graph) nuke()`

Do we need to export the `Root` property? I feel like that should be hidden.

dmcgowan replied:
`nukeGraph` could certainly be renamed

As for `Root`, the approach I took in this refactor was to leave any variable or method exposed which is called by the push or pull code. This will make moving that code out easier. `Root` is used by the push flow here https://github.com/docker/docker/blob/master/graph/push.go#L251, although the best approach is probably to change that usage and hide the `Root` property.

dmcgowan replied:
just saw this was in a test file, defining a helper function makes more sense than defining a method in that context

stevvooe replied:
O, sorry, didn't see file location.

dmcgowan replied:
Actually for `Root`, that case in push should just be calling `s.graph.RawJSON`, and not that it takes an ID instead of an image the signature matches :smile:. I will go ahead and unexport it.

stevvooe commented on graph/history.go:
```diff
@@ -8,6 +9,78 @@ import (
 )
 
+// History returns the list of all images used to create this image.
+func (graph *Graph) History(img *image.Image) ([]*image.Image, error) {
```
Not sure if this is relevant: what is the advantage of the pointer types for return and `WalkHistory` callbacks? It feels like we are exposing unprotected memory without proper locking. Yes, it is less memory efficient, but that may be a premature optimization. The output of history is small and the callback would nearly be all on stack if you're careful.

dmcgowan replied:
The `*image.Image` is not altered after creation. Obviously there is nothing to protect it, but that pattern is used throughout the entire package. Would be better to protect with an interface than just change a large struct to pass by value.

The `History` is not actually called anywhere, I only kept it because it appears to have a clear purpose.

stevvooe replied:
@dmcgowan Do not use efficiency to make by-value vs by-reference decisions. It generally has no effect on performance, especially in applications like this. It is better to copy here and maintain data integrity.

dmcgowan replied:
Pushed an extra commit which removes `History` and makes updates the WalkHistory passed in function to pass by value. It is only used in two cases and the pointer reference is held is neither. As I pointed out before, `History` was never actually called.

dmcgowan committed:
Separate graph from image

Move graph related functions in image to graph package.
Consolidating graph functionality is the first step in refactoring graph into an image store model.
Subsequent refactors will involve breaking up graph into multiple types with a strongly defined interface.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan committed:
Update graph walkhistory to pass by value

Remove unused graph history function

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan commented:
@stevvooe added a commit which makes the get/set checksum methods on graph use `digest.Digest` instead

stevvooe commented on graph/graph.go:
```diff
@@ -428,10 +435,129 @@ func (graph *Graph) Heads() (map[string]*image.Image, error) {
+	return nil
+}
+
+func (graph *Graph) SetCheckSum(id, cs string) error {
```
This looks a little dangerous...

dmcgowan replied:
Oops forgot to remove, had it here to check back references with oracle

dmcgowan replied:
Removed!

stevvooe commented on graph/graph.go:
```diff
@@ -428,10 +435,129 @@ func (graph *Graph) Heads() (map[string]*image.Image, error) {
+	return nil
+}
+
+// GetCheckSum gets the checksum for the provide image layer id.
```
Let's note that this is the digest of the tar layer and not the manifest associated with this image.

dmcgowan replied:
Yes, I anticipate this function to be moved to a LayerStore and this will never represent a digest which refers to a manifest.

dmcgowan committed:
Update graph to use digest type

Update get and set functions to use digests.
Update push code to use the digest type instead of string

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

stevvooe commented:
LGTM

icecrime commented:
Yes :+1: LGTM
